tests/foreman/ui/test_hostcollection.py

```diff
@@ -383,3 +404,223 @@ def test_negative_hosts_limit(self):
+            'repository-set': REPOSET['rhst7'],
+            'repository': REPOS['rhst7']['name'],
+            'organization-id': self.org_.id,
+            'content-view-id': self.content_view.id,
+            'lifecycle-environment-id': self.env.id,
+            'activationkey-id': self.activation_key.id,
+        })
+        setup_org_for_a_custom_repo({
+            'url': FAKE_0_YUM_REPO,
+            'organization-id': self.org_.id,
+            'content-view-id': self.content_view.id,
+            'lifecycle-environment-id': self.env.id,
+            'activationkey-id': self.activation_key.id,
+        })
+        self.hosts = []
+        for _ in range(self.hosts_number):
+            client = VirtualMachine(distro='rhel71')
+            self.hosts.append(client)
+            client.create()
+            client.install_katello_ca()
+            result = client.register_contenthost(
+                self.org_.label, self.activation_key.name)
+            self.assertEqual(result.return_code, 0)
+            client.enable_repo(REPOS['rhst7']['id'])
+            client.install_katello_agent()
+        host_ids = [
+            entities.Host().search(query={
+                'search': 'name={0}'.format(host.hostname)})[0].id
+            for host in self.hosts
+        ]
+        self.host_collection = entities.HostCollection(
+            host=host_ids,
+            organization=self.org_,
+        ).create()
+
+    def tearDown(self):
+        """Destroy all the VMs"""
+        for client in self.hosts:
+            client.destroy()
+        super(HostCollectionPackageManagementTest, self).tearDown()
+
+    def _validate_package_installed(self, hosts, package_name,
+                                    expected_installed=True, timeout=120):
+        """Check whether package was installed on the list of hosts."""
+
+        for host in hosts:
+            for _ in range(timeout / 15):
+                result = self.contenthost.package_search(
+                    host.hostname, package_name)
+                if (result is not None and expected_installed or
+                        result is None and not expected_installed):
+                    break
+                sleep(15)
+            else:
+                self.fail(
+                    u'Package {0} was not {1} host {2}'.format(
+                        package_name,
+                        'installed on ' if expected_installed else
+                        'removed from ',
+                        host.hostname,
+                    )
+                )
+
+    @tier3
+    def test_positive_install_package(self):
+        """Install a package to hosts inside host collection remotely
+
+        @id: eead8392-0ffc-4062-b045-5d0252670775
+
+        @assert: Package was successfully installed on all the hosts in host
+        collection
+
+        @CaseLevel: System
+        """
+        with Session(self.browser) as session:
+            session.nav.go_to_select_org(self.org_.name)
```

Comment:
@abalakh what you said is true when we use only `-n x` but if we put `--boxed` then, each `test_` function run in a separate process, having module and setupClass executed again.

I checked it

```python
import random
import logging
from unittest2 import TestCase

logger = logging.getLogger()
formatter = logging.Formatter(
    fmt='%(asctime)s - %(name)s - %(levelname)s - %(message)s',
    datefmt='%Y-%m-%d %H:%M:%S'
)
handler = logging.FileHandler('test_boxed.log')
handler.setFormatter(formatter)
logger.addHandler(handler)
logger.setLevel('DEBUG')

            
class TestBoxed(TestCase):
    @classmethod
    def setUpClass(cls):
         cls.random = random.random()
         logger.info("Executing setupclass for %s", cls.random)
    
    @classmethod     
    def tearDownClass(cls):
         logger.info("Executing tearDownclass for %s" % cls.random)
         
    def test_one(self):
        logger.info( "Executing test 1 for %s" % self.random)
        
    def test_two(self):
        logger.info( "Executing test 2 for %s" % self.random)
        
    def test_three(self):
        logger.info( "Executing test 3 for %s" % self.random)
        
    def test_four(self):
        logger.info( "Executing test 4 for %s" % self.random)
        
    def test_five(self):
        logger.info( "Executing test 5 for %s" % self.random)
        
    def test_six(self):
        logger.info( "Executing test 6 for %s" % self.random)
        
    def test_seven(self):
        logger.info( "Executing test 7 for %s" % self.random)
        
    def test_eight(self):
        logger.info( "Executing test 8 for %s" % self.random)
```        
 
 Using  `-n 4` setUpClass, tearDownClass executed 4 times (one per thread) 
     
```bash
2016-07-19 11:22:53 - root - INFO - Executing setupclass for 0.253839755338
2016-07-19 11:22:53 - root - INFO - Executing setupclass for 0.0742128890458
2016-07-19 11:22:53 - root - INFO - Executing setupclass for 0.15721937884
2016-07-19 11:22:53 - root - INFO - Executing test 1 for 0.0742128890458
2016-07-19 11:22:53 - root - INFO - Executing test 8 for 0.253839755338
2016-07-19 11:22:53 - root - INFO - Executing test 4 for 0.15721937884
2016-07-19 11:22:53 - root - INFO - Executing test 2 for 0.0742128890458
2016-07-19 11:22:53 - root - INFO - Executing test 7 for 0.253839755338
2016-07-19 11:22:53 - root - INFO - Executing tearDownclass for 0.0742128890458
2016-07-19 11:22:53 - root - INFO - Executing tearDownclass for 0.253839755338
2016-07-19 11:22:53 - root - INFO - Executing test 3 for 0.15721937884
2016-07-19 11:22:53 - root - INFO - Executing tearDownclass for 0.15721937884
2016-07-19 11:22:53 - root - INFO - Executing setupclass for 0.68192226145
2016-07-19 11:22:53 - root - INFO - Executing test 5 for 0.68192226145
2016-07-19 11:22:53 - root - INFO - Executing test 6 for 0.68192226145
2016-07-19 11:22:53 - root - INFO - Executing tearDownclass for 0.68192226145
```

 Using  `-n 4 --boxed` setUpClass, tearDownClass executed 8 times (one per test function) 

```
2016-07-19 11:23:27 - root - INFO - Executing setupclass for 0.548204210448
2016-07-19 11:23:27 - root - INFO - Executing setupclass for 0.911763900663
2016-07-19 11:23:27 - root - INFO - Executing setupclass for 0.660116234531
2016-07-19 11:23:27 - root - INFO - Executing test 8 for 0.548204210448
2016-07-19 11:23:27 - root - INFO - Executing tearDownclass for 0.548204210448
2016-07-19 11:23:27 - root - INFO - Executing test 5 for 0.911763900663
2016-07-19 11:23:27 - root - INFO - Executing tearDownclass for 0.911763900663
2016-07-19 11:23:27 - root - INFO - Executing test 1 for 0.660116234531
2016-07-19 11:23:27 - root - INFO - Executing setupclass for 0.338955669276
2016-07-19 11:23:27 - root - INFO - Executing tearDownclass for 0.660116234531
2016-07-19 11:23:27 - root - INFO - Executing test 4 for 0.338955669276
2016-07-19 11:23:27 - root - INFO - Executing tearDownclass for 0.338955669276
2016-07-19 11:23:27 - root - INFO - Executing setupclass for 0.548204210448
2016-07-19 11:23:27 - root - INFO - Executing setupclass for 0.911763900663
2016-07-19 11:23:27 - root - INFO - Executing test 7 for 0.548204210448
2016-07-19 11:23:27 - root - INFO - Executing tearDownclass for 0.548204210448
2016-07-19 11:23:27 - root - INFO - Executing test 6 for 0.911763900663
2016-07-19 11:23:27 - root - INFO - Executing tearDownclass for 0.911763900663
2016-07-19 11:23:27 - root - INFO - Executing setupclass for 0.660116234531
2016-07-19 11:23:27 - root - INFO - Executing test 2 for 0.660116234531
2016-07-19 11:23:27 - root - INFO - Executing tearDownclass for 0.660116234531
2016-07-19 11:23:27 - root - INFO - Executing setupclass for 0.338955669276
2016-07-19 11:23:27 - root - INFO - Executing test 3 for 0.338955669276
2016-07-19 11:23:27 - root - INFO - Executing tearDownclass for 0.338955669276

```


So we are running with --boxed, making setupclass and teardownclass isolated per test function